### DISCUSSION TITLE
Platform spec' in gem names

### DISCUSSION THREAD

**Discussion by wbreeze** [sparklemotion/nokogiri#3057] [General Q&A] [answered]
Hi! Thank you for this great gem. It's a gem. And my, it's widely used. Structured document parsing and search; XML. Close to my heart. Much appreciation to all of you.

Is it standard practice to put platform specifications in gem names? Can a gem do native install without being platform specific? What I'm seeing, developing on Mac and deploying on Linux, is some thrashing on source control of the file, `Gemfile.lock`. An example diff below. It's causing me to ignore diff's in `Gemfile.lock`, pass it over when doing commits. (More so than the heavy-handed "BUNDLED WITH".)

Thank you again for this gem of a gem.

Example:
```
diff --git a/Gemfile.lock b/Gemfile.lock
index 6768bb2..d7b1cd5 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -216,10 +216,10 @@ GEM
       jekyll-feed (~> 0.9)
       jekyll-seo-tag (~> 2.1)
     minitest (5.20.0)
-    nokogiri (1.15.4)
-      mini_portile2 (~> 2.8.2)
+    nokogiri (1.13.10)
+      mini_portile2 (~> 2.8.0)
       racc (~> 1.4)
-    nokogiri (1.15.4-x86_64-linux)
+    nokogiri (1.13.10-x86_64-darwin)
       racc (~> 1.4)
     nokogiri (1.13.10-x86_64-linux)
       racc (~> 1.4)
@@ -272,4 +272,4 @@ DEPENDENCIES
   jekyll-docs
 
 BUNDLED WITH
-   2.3.26
+   2.4.19
```

**Comment by flavorjones** [comment, marked as answer]
Hi @wbreeze, thanks for the kind words and for asking this question.

Nokogiri ships precompiled native gems, and that's why the platform name is showing up in your lock file. This isn't commonly done -- most C extensions are compiled from source during installation.

The first and most important piece of advice I have is: make sure you are on the latest version of bundler on all your platforms. The behavior around resolving platform-specific versions of gems has improved a lot over the last two years. Being on the latest version means that some of the bugs that have been fixed will no longer be present for you.

The other thing you should do is to `bundle lock --add-platform <platform-name>` for all the platforms you need your project to work on. For example, if you're developing on mac (and members of your team have both intel and apple silicon) and deploying on linux x86-64, you should make sure to run:

```
bundle lock --add-platform x86_64-darwin
bundle lock --add-platform arm64-darwin
bundle lock --add-platform x86_64-linux
```

and then your Gemfile.lock will have all the necessary dependencies for you to be able to `bundle install` on those three platforms.

The [Installing Native Gems](https://nokogiri.org/tutorials/installing_nokogiri.html#installing-native-gems) section in the docs has more information.

Hope this helps!
